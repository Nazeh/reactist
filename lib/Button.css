.reactist_button {
  font-size: 0.875rem;
  color: white;
  font-weight: 400;
  line-height: 1.7;
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  box-sizing: border-box;
  background-color: #3f82ef;
  padding: 5px 15px;
  cursor: pointer;
}
.reactist_button .wrapper {
  display: inline-block;
}
.reactist_button span {
  user-select: none;
}
.reactist_button:disabled,
.reactist_button.busy {
  opacity: 0.4;
  cursor: not-allowed;
}
.reactist_button:not(.secondary):not(.white):not(.danger):hover {
  background-color: #3b7be2;
}
.reactist_button.small {
  font-size: 0.81rem;
  color: white;
  font-weight: 400;
  line-height: 1.6;
  padding: 5px 10px;
}
.reactist_button.large {
  padding: 10px 15px;
}
.reactist_button.danger {
  background-color: #de4c4a;
}
.reactist_button.danger:not([disabled]):not(.busy):hover {
  background-color: #cf2826;
}
.reactist_button.secondary {
  background-color: white;
  border: 1px solid #dcdcdc;
  border-radius: 3px;
  color: #202020;
}
.reactist_button.secondary:not([disabled]):not(.busy):hover {
  background-color: #f9f9f9;
}
.reactist_button.white {
  background-color: white;
  border: 1px solid #3f82ef;
  border-radius: 3px;
  color: #3f82ef;
}
.reactist_button.white:not([disabled]):not(.busy):hover {
  background-color: #e7f3fd;
}
.reactist_button.busy {
  cursor: progress !important;
}
.reactist_button.busy.white,
.reactist_button.busy.secondary {
  border: 1px solid #dcdcdc;
  border-radius: 3px;
  background-color: #dcdcdc;
  color: #808080;
}
.reactist_button.busy:after {
  background-repeat: no-repeat;
  background-size: 15px;
  content: '';
  display: inline-block;
  height: 15px;
  margin-left: 12px;
  vertical-align: middle;
  width: 15px;
  animation-duration: 1s;
  animation-iteration-count: infinite;
  animation-name: reactistRotateRight;
  animation-timing-function: linear;
  background-image: url("data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' width='16.257px' height='16.257px' viewBox='-147.813 206.75 16.257 16.257' enable-background='new -147.813 206.75 16.257 16.257' xml:space='preserve'%3E %3Cg id='Symbols'%3E %3Cg id='Loading-spinner' transform='translate(0.000000, -2.000000)'%3E %3Cg id='Group'%3E %3Cg id='Rectangle-12'%3E %3C/g%3E %3Cdefs%3E %3Cfilter id='Adobe_OpacityMaskFilter' filterUnits='userSpaceOnUse' x='-147.684' y='210.929' width='16' height='13.95'%3E %3CfeColorMatrix type='matrix' values='1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0'/%3E %3C/filter%3E %3C/defs%3E %3Cmask maskUnits='userSpaceOnUse' x='-147.684' y='210.929' width='16' height='13.95' id='mask-2_1_'%3E %3Cg filter='url(%23Adobe_OpacityMaskFilter)'%3E %3Crect id='path-1_1_' x='-148.584' y='207.979' fill='%23FFFFFF' width='18' height='18'/%3E %3C/g%3E %3C/mask%3E %3Cpath id='Oval' mask='url(%23mask-2_1_)' fill='none' stroke='%23FFFFFF' stroke-width='2' stroke-linecap='round' stroke-linejoin='round' d=' M-144.634,211.929c-2.734,2.734-2.734,7.166,0,9.899l0,0c2.734,2.734,7.166,2.734,9.899,0c2.734-2.734,2.734-7.166,0-9.899'/%3E %3C/g%3E %3C/g%3E %3C/g%3E %3C/svg%3E");
}
@keyframes reactistRotateRight {
  from {
    transform: rotate(0deg);
    transform-origin: center center;
  }
  to {
    transform: rotate(360deg);
    transform-origin: center center;
  }
}

.reactist_tooltip__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_tooltip__wrapper .reactist_tooltip__text {
  font-size: 0.81rem;
  color: #202020;
  font-weight: 400;
  line-height: 1.6;
  text-align: center;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  padding: 5px 10px;
  max-width: 250px;
}
.reactist_tooltip__wrapper .reactist_tooltip__text:not(.inverted) {
  background-color: #333333;
  color: white;
  border: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:before {
  border-bottom-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:before {
  border-left-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:before {
  border-top-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:before {
  border-right-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:after {
  display: none;
}

.reactist_popover {
  white-space: normal;
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  display: none;
  visibility: hidden;
  opacity: 0;
  transition: opacity 0.1s linear;
}
.reactist_popover__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_popover .reactist_popover__content {
  max-width: 100%;
  display: inline-flex;
  border-radius: 3px;
  color: #202020;
  background-color: white;
  border: 1px solid #dcdcdc;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
}
.reactist_popover.visible {
  display: inline-block;
  display: initial;
  visibility: visible;
  opacity: 1;
}
.reactist_popover.arrow_top:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #dcdcdc;
}
.reactist_popover.arrow_top:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid white;
}
.reactist_popover.arrow_right:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_right:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -4px;
  border-left: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_bottom:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #dcdcdc;
}
.reactist_popover.arrow_bottom:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid white;
}
.reactist_popover.arrow_left:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_left:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -4px;
  border-right: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}

