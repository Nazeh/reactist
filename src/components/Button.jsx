import './styles/button.less'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Tooltip from './Tooltip'

class Button extends React.Component {
    constructor(props, context) {
        super(props, context)
        this._onClick = this._onClick.bind(this)
    }

    _onClick(event) {
        event.preventDefault()
        if (!this.props.disabled && this.props.onClick) {
            this.props.onClick()
        }
    }

    render() {
        const { data_tip } = this.props

        const className = classNames(
            {
                reactist: true,
                button: true,
                secondary: this.props.secondary,
                small: this.props.small,
                white: this.props.white,
                loading: this.props.loading,
                danger: this.props.danger
            },
            this.props.className
        )

        const button = (
            <button
                className={className}
                disabled={this.props.disabled}
                onClick={this._onClick}
            >
                <div className="wrapper">
                    <span>{this.props.name}</span>
                </div>
            </button>
        )

        // conditionally wrap into tooltip
        return data_tip ? <Tooltip text={data_tip}>{button}</Tooltip> : button
    }
}
Button.displayName = 'Button'
Button.defaultProps = {
    secondary: false,
    small: false,
    white: false,
    loading: false,
    disabled: false,
    danger: false
}
Button.propTypes = {
    /** Text that is displayed on the button. */
    name: PropTypes.string,
    /**
     * Function that is called when the button is clicked.
     * Is only invoked when disabled is not set.
     */
    onClick: PropTypes.func,
    /** Secondary style. */
    secondary: PropTypes.bool,
    /** Small style. */
    small: PropTypes.bool,
    /** White style. */
    white: PropTypes.bool,
    /** Loading style. */
    loading: PropTypes.bool,
    /** Disabled style. Prevents onClick from being called. */
    disabled: PropTypes.bool,
    /** Danger style. */
    danger: PropTypes.bool,
    /** Tooltip that is displayed on hover. */
    data_tip: PropTypes.string,
    /** Additional css class applied to the button. */
    className: PropTypes.string
}

export default Button
