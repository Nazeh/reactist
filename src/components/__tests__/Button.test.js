import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Button from '../Button.jsx'

describe('Button', () => {
    it('renders empty without crashing', () => {
        const button = shallow(<Button />)
        expect(toJson(button)).toMatchSnapshot()
    })

    it('renders name as button text', () => {
        const button = shallow(<Button name="THE BUTTON TEXT" />)
        expect(toJson(button)).toMatchSnapshot()
    })

    it('calls onclick when clicked', () => {
        const mockEvent = { preventDefault: jest.fn() }
        const clickSpy = jest.fn()
        const button = shallow(<Button onClick={clickSpy} />)

        button.find('button').simulate('click', mockEvent)
        expect(clickSpy).toHaveBeenCalled()
    })

    it('does not call onclick when disabled and clicked', () => {
        const mockEvent = { preventDefault: jest.fn() }
        const clickSpy = jest.fn()
        const button = shallow(<Button disabled onClick={clickSpy} />)

        button.find('button').simulate('click', mockEvent)
        expect(clickSpy).not.toHaveBeenCalled()
    })

    it('does not call onClick when loading and clicked', () => {
        const mockEvent = { preventDefault: jest.fn() }
        const clickSpy = jest.fn()
        const button = shallow(<Button loading onClick={clickSpy} />)

        button.find('button').simulate('click', mockEvent)
        expect(clickSpy).not.toHaveBeenCalled()
    })

    it('renders a tooltip when prop is supplied', () => {
        const button = shallow(
            <Button name="THE BUTTON TEXT" data_tip="THE TOOLTIP" />
        )
        expect(toJson(button)).toMatchSnapshot()
    })

    it('adds additional className when supplied', () => {
        const button = shallow(
            <Button
                name="THE BUTTON TEXT"
                className="very-complex classnames-are-added"
            />
        )
        expect(toJson(button)).toMatchSnapshot()
    })
})
