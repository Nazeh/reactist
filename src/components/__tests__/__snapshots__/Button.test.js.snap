// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button adds additional className when supplied 1`] = `
<button
  className="reactist button very-complex classnames-are-added"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="wrapper"
  >
    <span>
      THE BUTTON TEXT
    </span>
  </div>
</button>
`;

exports[`Button renders a tooltip when prop is supplied 1`] = `
<Tooltip
  allowVaguePositioning={false}
  delayHide={0}
  delayShow={1000}
  hideOnScroll={true}
  position="auto"
  text="THE TOOLTIP"
>
  <button
    className="reactist button"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="wrapper"
    >
      <span>
        THE BUTTON TEXT
      </span>
    </div>
  </button>
</Tooltip>
`;

exports[`Button renders empty without crashing 1`] = `
<button
  className="reactist button"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="wrapper"
  >
    <span />
  </div>
</button>
`;

exports[`Button renders name as button text 1`] = `
<button
  className="reactist button"
  disabled={false}
  onClick={[Function]}
>
  <div
    className="wrapper"
  >
    <span>
      THE BUTTON TEXT
    </span>
  </div>
</button>
`;
